### PR TITLE
readme: add section on package removal

### PR DESCRIPTION
**Pull request by ajeddeloh**
Document pitfalls and useful tricks for discovering where a package is
used.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,9 +20,46 @@ If you'd prefer to use a local copy of the portage tree, you can point
 Licensing information can be found in the respective files, so consult
 them directly. Most ebuilds are licensed under the GPL version 2.
 
+Upstream Gentoo sources: http://sources.gentoo.org/gentoo-x86/
+
+# Removing packages
+
+Be careful when removing packages. The following section offers tips for preventing
+breakage, but they are by no means exhaustive. Gentoo's portage utilities wont take
```
Comment (bgilbert) [inline README.md]: `won't`

File: README.md [hunk 2]
```diff
@@ -18,9 +20,46 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+into account some pretty core packages like catalyst or glibc, so take extra care
```
Comment (bgilbert) [inline README.md]: Is `package` the right way to describe the catalyst infrastructure?

File: README.md [hunk 3]
```diff
@@ -18,9 +20,46 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+
+`git log <category>/<package>` will show commits that touched that directory. These can give
+clues about why a package was included in the first place and where to look to ensure it really
+is unused.
+
+## equery d
+
+`equery d -a <package-name>` will tell you what packages depend on `package-name`. It will
+also generate a lot of false positives, since it considers all dependencies for all use flags,
+even ones we dont use, such as `test`.
```
Comment (bgilbert) [inline README.md]: `don't`

File: README.md [hunk 4]
```diff
@@ -18,9 +20,46 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+
+## emerge --emptytree
+`emerge --pretend --verbose --emptytree <package-name>` _should_ give a list of all the dependencies
+for a given package. After removing the ebuild and package, make sure it's removal doesn't break
```
Comment (bgilbert) [inline README.md]: `its`

File: README.md [hunk 5]
```diff
@@ -18,9 +20,46 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+the packages `board-packages` and `sdk-depends`. Remember to use the `emerge-<arch>-usr` commands
+to check `board-packages` and `emerge` to check `sdk-depends`.
+
+The ebuild must be removed and the package must be removed locally or the pacakge
```
Comment (bgilbert) [inline README.md]: `package`

File: README.md [hunk 6]
```diff
@@ -18,9 +20,46 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+will be silently elided in package list. To see if there are any packages installed
```
Comment (bgilbert) [inline README.md]: `elided from the`

File: README.md (hunk 2 above)
Comment (ajeddeloh) [reply README.md]: Probably not. It's not clear to me how exactly it's used. We have a catalyst package (`dev-util/catalyst`) but I can't find a single place we install it or use it.

Comment (bgilbert) [reply README.md]: `scripts/build_library/catalyst.sh`

Comment (ajeddeloh) [reply README.md]: After talking with @dm0-, jenkins installs this. I'm adding it to `sdk-extras` (https://github.com/coreos/coreos-overlay/pull/2867) so we can track it and its dependencies.

**Comment by ajeddeloh** [comment]
Fixed typos and expanded contractions, made some of the language more precise.

File: README.md [hunk 7]
```diff
@@ -18,9 +22,49 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+
+## emerge --emptytree
+
+`emerge --pretend --verbose --emptytree <package-name>` _should_ give a list of all the dependencies
+for a given package. After removing the ebuild and package, make sure its removal doesn't break
+the packages `board-packages` and `sdk-depends`. Remember to use the `emerge-<arch>-usr` commands
+to check `board-packages` and `emerge` to check `sdk-depends`.
+
+The ebuild must be removed from `portage-stable` and the package must be removed locally or else
+the package will be silently elided in the dependency list. To see if there are any packages
```
Comment (bgilbert) [inline README.md]: This appears to say that if I leave the ebuild in place and leave the package installed, it won't show up in the dep list, which doesn't sound right. Is the intent to say that the problem occurs if I remove the ebuild but leave the package installed?

**Comment by ajeddeloh** [comment]
Updated:
 - Language on emptytree behaviour
 - Fixed `egencache` command and monospaced it
 - Added `@system` to the list of things to check

Added:
 - Note about emptytree and virtuals
 - Note about running tests on Jenkins
 - Note about `update_metadata` script
 - Note about checking `coreos-overlay` for things like use flags or accept_keywords.

File: README.md [hunk 8]
```diff
@@ -18,9 +22,64 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+```
+    egencache --update --repo portage-stable
+```
+There is also the `update_metadata` script which will update both `portage-stable` and `coreos-overlay`
+and optionally generate a commit message.
 
-    egencache --update portage-stable
-
-Upstream Gentoo sources: http://sources.gentoo.org/gentoo-x86/
+## Testing changes
```
Comment (bgilbert) [inline README.md]: nit: blank line

File: README.md [hunk 9]
```diff
@@ -18,9 +22,64 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+If you work at CoreOS, or have Jenkins running with our configuration, you can make pull requests with your changes and
```
Comment (bgilbert) [inline README.md]: The first part isn't necessary. `If you're running the [jenkins-os](https://github.com/coreos/jenkins-os) Jenkins configuration, ...`

**Comment by ajeddeloh** [comment]
Added one more little bit when talking about emptytree to mention packages needed by the bootstrapping process.

File: README.md [hunk 10]
```diff
@@ -18,9 +22,72 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+
+`equery d -a <package-name>` will tell you what packages depend on `package-name`. It will
+also generate a lot of false positives, since it considers all dependencies for all use flags,
+even ones we do not use, such as `test`.
+
+## emerge --emptytree
+
+`emerge --pretend --verbose --emptytree <package-name>` _should_ give a list of all the dependencies
+for a given package. Use this to test if `board-packages`, `sdk-depends`, and `@system` can still be
+emerged after removing an ebuild and pacakge.
```
Comment (bgilbert) [inline README.md]: `package`

File: README.md [hunk 11]
```diff
@@ -18,9 +22,72 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+Remember to use the `emerge-<arch>-usr` commands to check `board-packages` and `emerge` to check
+`sdk-depends`. Use both when checking `@system`.
+
+Furthermore, the SDK bootstrapping process uses a list of packages defined the by SDK profile's packages.build
+file. Install dev-util/catalyst and run /usr/lib64/catalyst/targets/stage1/build.py to get a list of packages
```
Comment (bgilbert) [inline README.md]: Missing backticks.

File: README.md [hunk 12]
```diff
@@ -18,9 +22,72 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+needed for the boostrapping process, then run `emerge --emptytree` on that list.
+
+The ebuild must be removed from `portage-stable` _and_ the package must be removed locally. If only the
```
Comment (bgilbert) [inline README.md]: `A package's ebuild` would better introduce the subject change.

File: README.md [hunk 13]
```diff
@@ -18,9 +22,72 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+ebuild is removed, the package will be silently elided in the `emerge --emptytree` dependency list.
+To see if there are any packages installed without ebuilds run `eix -tTc`. There are no `eix-<arch>-usr` wrappers, so double
+check the packages are also unmerged via the `emerge-<arch>-usr` commands. Make sure to run
+`eix-update` before running other `eix` commands.
+
+`emerge --emptytree` also has unintuitive behavior when handling virtual packages.
+When making changes affecting virtual packages (removing a provider, changing use flags that affect
+a provider or virtual, etc), `emerge --emptytree` will always prefer an already installed
+provider (unless it is masked or otherwise disabled), so unmerge the current provider before
+running `emerge --emptytree` to ensure the virtuals are resolved correctly. Look at the virtual packages's
```
Comment (bgilbert) [inline README.md]: `package's`

File: README.md [hunk 14]
```diff
@@ -18,9 +22,72 @@ If you'd prefer to use a local copy of the portage tree, you can point
...
+    egencache --update --repo portage-stable
+```
+There is also `scripts/update_metadata` which will update both `portage-stable` and `coreos-overlay`
+and optionally generate a commit message.
 
-    egencache --update portage-stable
+## Testing changes
 
-Upstream Gentoo sources: http://sources.gentoo.org/gentoo-x86/
+If you have Jenkins running with our configuration, you can make pull requests with your changes and
```
Comment (bgilbert) [inline README.md]: This is okay, but `our configuration` seems unnecessarily possessive. IMO it reads better to just name the config repo and link to it.

**Pull request merged**